File: src/log-buffer.ts
const MAX_LINES_PER_PROCESS = 1000;

export interface LogEntry {
  line: string;
  source: 'stdout' | 'stderr';
  timestamp: number;
  processId?: number;
}

export class LogBuffer {
  private buffers: Map<number, LogEntry[]>;
  private unifiedBuffer: LogEntry[];
  private maxLines: number;

  constructor() {
    this.buffers = new Map();
    this.unifiedBuffer = [];
    this.maxLines = MAX_LINES_PER_PROCESS;
  }

  addLog(processId: number, line: string, source: 'stdout' | 'stderr' = 'stdout'): void {
    if (!this.buffers.has(processId)) {
      this.buffers.set(processId, []);
    }

    const buffer = this.buffers.get(processId)!;
    buffer.push({ line, source, timestamp: Date.now() });

    if (buffer.length > this.maxLines) {
      buffer.shift();
    }

    this.unifiedBuffer.push({ processId, line, source, timestamp: Date.now() });
    if (this.unifiedBuffer.length > this.maxLines * 10) {
      this.unifiedBuffer.shift();
    }
  }

  getLogs(processId: number): LogEntry[] {
    return this.buffers.get(processId) || [];
  }

  getUnifiedLogs(): LogEntry[] {
    return this.unifiedBuffer;
  }

  clear(processId?: number): void {
    if (processId !== undefined) {
      this.buffers.delete(processId);
    } else {
      this.buffers.clear();
      this.unifiedBuffer = [];
    }
  }
}

File: log_buffer.go
package main

import (
	"slices"
	"sync"
	"time"
)

const maxLinesPerProcess = 1000

type LogEntry struct {
	Line      string
	Source    LogSource
	Timestamp time.Time
	ProcessID int
	HasID     bool
	IsSystem  bool
}

type LogBuffer struct {
	mu      sync.RWMutex
	buffers map[int][]LogEntry
	unified []LogEntry
	max     int
}

func NewLogBuffer() *LogBuffer {
	return &LogBuffer{
		buffers: make(map[int][]LogEntry),
		max:     maxLinesPerProcess,
	}
}

func (l *LogBuffer) Add(processID int, line string, source LogSource, isSystem bool) {
	l.mu.Lock()
	defer l.mu.Unlock()

	entry := LogEntry{
		Line:      line,
		Source:    source,
		Timestamp: time.Now(),
		ProcessID: processID,
		HasID:     true,
		IsSystem:  isSystem,
	}

	l.buffers[processID] = append(l.buffers[processID], entry)
	if len(l.buffers[processID]) > l.max {
		l.buffers[processID] = l.buffers[processID][1:]
	}

	l.unified = append(l.unified, entry)
	if len(l.unified) > l.max*10 {
		l.unified = l.unified[1:]
	}
}

func (l *LogBuffer) Logs(processID int) []LogEntry {
	l.mu.RLock()
	defer l.mu.RUnlock()
	return append([]LogEntry(nil), l.buffers[processID]...)
}

func (l *LogBuffer) UnifiedLogs() []LogEntry {
	l.mu.RLock()
	defer l.mu.RUnlock()
	return append([]LogEntry(nil), l.unified...)
}

func (l *LogBuffer) LogsForProcessIDs(ids []int) []LogEntry {
	l.mu.RLock()
	defer l.mu.RUnlock()

	merged := make([]LogEntry, 0)
	for _, id := range ids {
		merged = append(merged, l.buffers[id]...)
	}
	slices.SortFunc(merged, func(a, b LogEntry) int {
		return a.Timestamp.Compare(b.Timestamp)
	})
	return merged
}

func (l *LogBuffer) Clear(processID *int) {
	l.mu.Lock()
	defer l.mu.Unlock()
	if processID != nil {
		delete(l.buffers, *processID)
		return
	}
	l.buffers = make(map[int][]LogEntry)
	l.unified = nil
}
